### PR TITLE
Refactoring and visualization

### PR DESCRIPTION
**Pull request by bilabbasi**
The main features added from this PR are as follows:

- a separate Metric class to simplify the FaceMetric, TangentMetric, etc.
- a pooling operation for meshes
- adds scripts to easily run experiments
- adds scripts for visualization of meshes
- the train scripts have been significantly refactors for ease of reuse through out the repo (and not to be used as just a stand alone script)

### CODE REVIEW

**Comment by trachelr** [comment]
Hi Bilal, thanks a lot for your PR.
Let me test it and do a quick review.

File: README.md [hunk 1]
```diff
@@ -67,26 +70,45 @@ out = net(features,mesh.pos,mesh.edge_index,mesh.face.t())
...
+
+* *FAUST.* For the correspondence task, we used the Fine Alignment Using Scan Texture ([FAUST](http://faust.is.tue.mpg.de/)) dataset. We rely on the [FAUST dataset module](https://github.com/rusty1s/pytorch_geometric/blob/master/torch_geometric/datasets/faust.py) in *PyTorch Geometric* for handling the dataset.
+Download the dataset from [here](http://faust.is.tue.mpg.de/) and move it to `data/FAUST/raw` (which may need to be manually created).
 
-For the segmentation task, we use the [Shape COSEG](http://irc.cs.sdu.edu.cn/~yunhai/public_html/ssl/ssd.htm) dataset and [Human Body Segmentation](https://github.com/pixelite1201/pytorch_coma).
+* *COSEG.* For the segmentation task, we use the [Shape COSEG](http://irc.cs.sdu.edu.cn/~yunhai/public_html/ssl/ssd.htm) dataset.To download and install the COSEG dataset, run the following command, which will download, split, and process labels of the COSEG dataset (respectively):
+    ```
+    python preprocess/install_coseg.py --data-dir data/COSEG --ratio 0.85
+    ```
+To use the exact train/test split used in our experiments, add the flag `--splits preprocess/coseg/splits.yaml` to the above command.
```
Comment (trachelr) [inline README.md:88]: could you specify that's the default value in the splits argument please?

File: environment.yml [hunk 2]
```diff
@@ -7,11 +7,17 @@ dependencies:
     - python==3.8
     - numpy==1.19.1
     - matplotlib==3.3.0
-    - pytorch==1.5.0
+    - pytorch
```
Comment (trachelr) [inline environment.yml:10]: I think pytorch_geometric is missing, please see pyg package here
https://pytorch-geometric.readthedocs.io/en/latest/notes/installation.html

File: models/mesh_pool.py [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+        Returns None or nn.Module (MLP) for score computation.
+
+        :param in_channels: Number of input channels per item
+
+        :return: None or MLP used to extract vertex/edge scores
+        """
+        if in_channels == 0:
+            return None
+        else:
+            return nn.Sequential(Linear(2, 1), nn.LeakyReLU())
```
Comment (trachelr) [inline models/mesh_pool.py]: `nn` is not defined, please use `torch.nn` here

File: scripts/visualize_metric.py [hunk 4]
```diff
@@ -1,60 +1,121 @@
...
+    v = (v - v.min()) / (v.max() - v.min())
+    lambda_min = lambda_max = torch.ones(len(model.metric_per_vertex), len(v))
+    for i, m in enumerate(model.metric_per_vertex):
+        weights = torch.ones(len(v))
+        for j in range(len(v)):
+            metric = m[j].t().mm(m[j])
+            eigs = metric.eig()[0][:, 0]
+            lambda_max[i, j] = eigs.max()
+            lambda_min[i, j] = eigs.min()
+            # weights[j] = (lambda_max*lambda_min)
```
Comment (trachelr) [inline scripts/visualize_metric.py]: please remove unnecessary comment

File: scripts/visualize_pools.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+            v = data.pos
+            f = data.face.t()
+            e = data.edge_index
+        else:
+            (v, e, f) = ds[i][0]
+        model(v, v, e, f)
+
+        j = 0
+        for (x, v, e, f, cluster) in model.mesh_per_layer:
+            # _,v,e,f,__ = pool(torch.rand(len(v)),v,e,f.type(torch.long).t())
```
Comment (trachelr) [inline scripts/visualize_pools.py]: same here, please remove these comments if it's not needed anymore...

File: preprocess/coseg/convert_labels.py [hunk 6]
```diff
@@ -1,45 +1,55 @@
...
+import os
 from os.path import join
 
-import torch
 import numpy as np
+import torch
 import trimesh
+import yaml
+
+config = yaml.safe_load(open("config.yml"))
```
Comment (trachelr) [inline preprocess/coseg/convert_labels.py:10]: please provide full path to `config.yml` or remove this line

File: preprocess/install_human_seg.py [hunk 7]
```diff
@@ -1,122 +1,33 @@
...
 
-import torch
-import trimesh
-import numpy as np
+import human_seg
+import yaml
 
-from psbody.mesh import Mesh
-import tmp
+# config = yaml.safe_load(open("config.yml"))
```
Comment (trachelr) [inline preprocess/install_human_seg.py:8]: please remove this line

File: tests/test_compute_face_area_and_angle.py [hunk 8]
```diff
@@ -79,6 +95,9 @@ def test_compute_face_area_and_angle():
...
     )
 
-    area, angle = metric_conv.compute_face_area_and_angle(pos, faces)
+    area, std, angle = metric.compute_face_area_and_angle(pos, faces)
 
-    assert (torch.abs(answer_area - area) < 1e-6).sum() == 4 and (torch.abs(answer_angle - angle) < 1e-6).sum() == 4
+    assert (torch.abs(answer_area - area) < 1e-6).sum() == 4
+    assert (torch.abs(answer_std - std) < 1e-4).sum() == 4
+    assert (torch.abs(answer_angle - angle) < 1e-6).sum() == 4
+    # assert (torch.abs(answer_area - area) < 1e-6).sum() == 4 and (torch.abs(answer_std-std)<1e-6).sum() == 4  and (torch.abs(answer_angle - angle) < 1e-6).sum() == 4
```
Comment (trachelr) [inline tests/test_compute_face_area_and_angle.py]: same here please

File: train/train_args.py [hunk 9]
```diff
@@ -1,40 +1,118 @@
...
-    parser.add_argument("--remove-reference",action='store_true', help="Removes reference node from message-passing.")
-    parser.add_argument("--symmetric",action='store_true',help="Uses (A+A^T)/2 as message-passing.")
-   
+    parser.add_argument("--model", type=str, required=False, help="Model to be trained.")
+    parser.add_argument(
+        "--classification",
+        action="store_true",
+        help="Indicate whether task is a classification task or not.",
+    )
+    parser.add_argument("layers", nargs="*", default=[16, 32, 64, 128], help="Architecture for MetricConvNet")
```
Comment (trachelr) [inline train/train_args.py]: missing `--` prefix in "layers" here

File: scripts/visualize_metric.py [hunk 10]
```diff
@@ -1,60 +1,121 @@
...
+    print(pred.eq(gt).sum().float() / len(gt))
+
+    pred = pred / pred.max().float()
+    utils.mesh2ply(
+        to_numpy(v),
+        to_numpy(f),
+        pred,
+        fname=join(target_dir, f"{args['dataset']}_{data_idx}_pred.ply"),
+    )
+    pdb.set_trace()
```
Comment (trachelr) [inline scripts/visualize_metric.py]: Please remove this line as well as the pdb import

File: scripts/visualize_metric.py [hunk 11]
```diff
@@ -1,60 +1,121 @@
...
+        x = v.clone()
+
+    out = model(x, v, e, f)
+    v = (v - v.min()) / (v.max() - v.min())
+    lambda_min = lambda_max = torch.ones(len(model.metric_per_vertex), len(v))
+    for i, m in enumerate(model.metric_per_vertex):
+        weights = torch.ones(len(v))
+        for j in range(len(v)):
+            metric = m[j].t().mm(m[j])
+            eigs = metric.eig()[0][:, 0]
```
Comment (trachelr) [inline scripts/visualize_metric.py:67]: There's a warning here, please use `torch.linalg.eig`

File: scripts/visualize_pools.py [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+
+
+def visualize_pools(
+    log_dir: str, data_dir: str, target_dir: str = None, num_samples: int = 1, train: bool = False
+) -> NoReturn:
+    if target_dir is None:
+        target_dir = join(log_dir, "samples")
+    os.makedirs(target_dir, exist_ok=True)
+
+    model, args = utils.get_model(log_dir, load_params=True)
```
Comment (trachelr) [inline scripts/visualize_pools.py]: could you check if the model is an `AutoEncoder` here?
I've got errors with other architectures because it doesnt have the `mesh_per_layer` attribute used below (L40)
Otherwise could you make this script compatible with other kind of models?

**Review by trachelr — COMMENTED** [review]
Very nice improvement of the code!! I was able to include pytorch_geometric in the `environment.yaml` file to install it with conda. I installed the CPU version in Ubuntu18.04 (through WSL) but I wasnt able to try GPU support. I've train a model for few epochs in the segmentation task with COSEG and HumanSeg datasets but I wasn't able to get the FAUST dataset for correspondance tasks. The scripts for metric visualisation is working on COSEG but I wasnt able to run the script for pools visualisation. Is it supposed to work with autoencoders only? There's a few pytests errors as well, please see logs in [test.txt](https://github.com/eidosmontreal/shape-analysis/files/8642015/test.txt).

Thanks again for your contribution!

File: environment.yml (hunk 2 above)
Comment (bilabbasi) [reply environment.yml:10]: The pytorch_geometric installation has separate instructions in the README, since it requires some additional stuff. There's a specific note in the README indicating that the environment.yaml doesn't contain installation for pytorch_geometric.

File: scripts/visualize_metric.py (hunk 11 above)
Comment (bilabbasi) [reply scripts/visualize_metric.py:67]: That's strange; I'm not getting an error. Just to make sure, is your PyTorch up-to-date? I just ran it on my side, and nothing.

**Comment by bilabbasi** [comment]
Hey Romain, thanks for the feedback. I've addressed a lot of the small things you've mentioned and submitted the changes. For the FAUST installation, can you tell me what your error is? If it persists, then I can just add details on how to download/set up the FAUST dataset.

I've removed the mesh pooling for now, since it's clear it requires a bit more testing before being implemented. 


Thanks for your review so far :)

File: environment.yml (hunk 2 above)
Comment (trachelr) [reply environment.yml:10]: my bad, I didn't read that note. I installed pyg version 2.0.4 and py38_torch_1.11.0_cu115
Let it like that if you'd like to keep the same installation version.

File: scripts/visualize_metric.py (hunk 11 above)
Comment (trachelr) [reply scripts/visualize_metric.py:67]: I have version 1.11

**Comment by trachelr** [comment]
Hello Bilal :)
Thanks for your changes and sorry for the mess-up with the installation... actually I'm running the code in a windows environment. For the FAUST dataset, it's a license issue, you know I'm not allowed to download the dataset... let's skip this part. Did you run the tests? I saw a few errors, don't know if your changes fixed it...
If you're confident with your changes, we can merge.

File: preprocess/coseg/convert_labels.py (hunk 6 above)
Comment (trachelr) [reply preprocess/coseg/convert_labels.py:10]: @bilabbasi did you saw this comment?
